File: python/bank-account/bank_account.py
import threading

class BankAccount(object):
    def __init__(self):
        self.balance = 0
        self.isOpen = False
        self.lock = threading.Lock()

    def get_balance(self):
        if not self.isOpen:
            raise ValueError('account is closed')
        return self.balance

    def open(self):
        if self.isOpen:
            raise ValueError('account already open')
        self.isOpen = True

    def deposit(self, amount):
        if not self.isOpen:
            raise ValueError('account is closed')
        if amount < 0:
            raise ValueError('negative deposit')
        
        with self.lock:
            self.balance += amount

    def withdraw(self, amount):
        if amount < 0:
            raise ValueError('negative withdrawal')
        if amount > self.balance:
            raise ValueError('overdrawn')
        
        with self.lock:
            self.balance -= amount

    def close(self):
        if not self.isOpen:
            raise ValueError('account already closed')
        
        with self.lock:
            self.isOpen = False
            self.balance = 0

File: java/bank-account/src/main/java/BankAccount.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

class BankAccount {
  private boolean isOpen = false;
  private double balance = 0;
  private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
  private final Lock writeLock = readWriteLock.writeLock();
  private final Lock readLock = readWriteLock.readLock();

  public double getBalance() throws BankAccountActionInvalidException {
    readLock.lock();
    double balanceToReturn;
    try {
      if (!isOpen) {
        throw new BankAccountActionInvalidException("Account closed");
      }
      balanceToReturn = balance;
    } finally {
      readLock.unlock();
    }
    return balanceToReturn;
  }

  void open() {
    writeLock.lock();
    try {
      isOpen = true;
    } finally {
      writeLock.unlock();
    }
  }

  void close() {
    writeLock.lock();
    try {
      isOpen = false;
    } finally {
      writeLock.unlock();
    }
  }

  void deposit(double amount) throws BankAccountActionInvalidException {
    if (amount <= 0) {
      throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
    }
    writeLock.lock();
    try {
      if (!isOpen) {
        throw new BankAccountActionInvalidException("Account closed");
      }
      balance += amount;
    } finally {
      writeLock.unlock();
    }
  }

  void withdraw(double amount) throws BankAccountActionInvalidException {
    if (amount <= 0) {
      throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
    }
    if (balance == 0) {
      throw new BankAccountActionInvalidException("Cannot withdraw money from an empty account");
    }
    if (amount > balance) {
      throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
    }
    writeLock.lock();
    try {
      if (!isOpen) {
        throw new BankAccountActionInvalidException("Account closed");
      }
      balance -= amount;
    } finally {
      writeLock.unlock();
    }
  }
}